apps/jupyter/extensions/notebook-runner/src/__tests__/notebook_runner.spec.ts

```diff
@@ -1,9 +1,300 @@
+        "grading data content",
+      );
+
+      const blob = await mockZip.generateAsync({ type: "blob" });
+
+      const result = await importCrtInternalTask(blob);
+
+      expect(result.taskTemplateFile).toBeInstanceOf(Blob);
+      expect(result.studentTaskFile).toBeInstanceOf(Blob);
+      expect(result.autograderFile).toBeInstanceOf(Blob);
+      expect(result.data).toBeDefined();
+      expect(result.gradingData).toBeDefined();
+      expect(result.src).toBeDefined();
+      expect(result.gradingSrc).toBeDefined();
+    });
+
+    it("should handle tasks with empty optional folders", async () => {
+      mockZip.file(CrtInternalFiles.Template, "template content");
+      mockZip.file(CrtInternalFiles.Student, "student content");
+      mockZip.file(CrtInternalFiles.Autograder, "autograder content");
+
+      const blob = await mockZip.generateAsync({ type: "blob" });
+
+      const result = await importCrtInternalTask(blob);
+
+      expect(result.data.size).toBe(0);
+      expect(result.gradingData.size).toBe(0);
+      expect(result.src.size).toBe(0);
+      expect(result.gradingSrc.size).toBe(0);
+    });
+
+    it("should correctly extract files from nested folders", async () => {
+      mockZip.file(CrtInternalFiles.Template, "template content");
+      mockZip.file(CrtInternalFiles.Student, "student content");
+      mockZip.file(CrtInternalFiles.Autograder, "autograder content");
+      mockZip.file(
+        `${CrtInternalFiles.Data}/nested/file.txt`,
+        "nested content",
+      );
+      mockZip.file(`${CrtInternalFiles.Src}/code.py`, "print('hello')");
+
+      const blob = await mockZip.generateAsync({ type: "blob" });
+
+      const result = await importCrtInternalTask(blob);
+
+      expect(result.data.get("nested/file.txt")).toBeInstanceOf(Blob);
+      expect(result.src.get("code.py")).toBeInstanceOf(Blob);
+    });
+
+    it("should handle binary files in data folder", async () => {
+      mockZip.file(CrtInternalFiles.Template, "template");
+      mockZip.file(CrtInternalFiles.Student, "student");
+      mockZip.file(CrtInternalFiles.Autograder, "autograder");
+
+      const binaryData = new Uint8Array([0x89, 0x50, 0x4e, 0x47]);
+      mockZip.file(`${CrtInternalFiles.Data}/image.png`, binaryData);
+
+      const blob = await mockZip.generateAsync({ type: "blob" });
+
+      const result = await importCrtInternalTask(blob);
+
+      expect(result.data.get("image.png")).toBeInstanceOf(Blob);
+    });
+  });
+
+  describe("missing required files", () => {
+    it("should throw MissingRequiredFilesError when template file is missing", async () => {
+      mockZip.file(CrtInternalFiles.Student, "student content");
+      mockZip.file(CrtInternalFiles.Autograder, "autograder content");
+
+      const blob = await mockZip.generateAsync({ type: "blob" });
+
+      await expect(importCrtInternalTask(blob)).rejects.toThrow(
+        MissingRequiredFilesError,
+      );
```

Comment:
Here and below: I would also assert the custom properties set on the error object (for instance: does it correctly detect which files are missing)